util/rowcodec/rowcodec_test.go

```diff
@@ -518,128 +450,109 @@ func (s *testSuite) TestTypesNewRowCodec(c *C) {
-				c.Assert(d.GetMysqlDecimal(), DeepEquals, t.bt.GetMysqlDecimal())
-			} else {
-				c.Assert(d, DeepEquals, t.bt)
+			// test encode input.
+			sc := new(stmtctx.StatementContext)
+			sc.TimeZone = time.UTC
+			newRow, err := encoder.Encode(sc, colIDs, dts, nil)
+			require.NoError(t, err)
+
+			// decode to datum map.
+			mDecoder := rowcodec.NewDatumMapDecoder(cols, sc.TimeZone)
+			dm, err := mDecoder.DecodeToDatumMap(newRow, nil)
+			require.NoError(t, err)
+			for _, td := range testCase.data {
+				d, exists := dm[td.id]
+				require.True(t, exists)
+				require.Equal(t, td.dt, d)
 			}
-		}
 
-		chk = chunk.New(fts, 1, 1)
-		cDecoder = rowcodec.NewChunkDecoder(cols, []int64{-1}, nil, sc.TimeZone)
-		err = cDecoder.DecodeToChunk(newRow, kv.IntHandle(-1), chk)
-		c.Assert(err, IsNil)
-		chkRow = chk.GetRow(0)
-		cdt = chkRow.GetDatumRow(fts)
-		for i := range testData {
-			if i == 0 {
-				continue
+			// decode to chunk.
+			cDecoder := rowcodec.NewChunkDecoder(cols, []int64{-1}, nil, sc.TimeZone)
+			chk := chunk.New(fts, 1, 1)
+			err = cDecoder.DecodeToChunk(newRow, kv.IntHandle(-1), chk)
+			require.NoError(t, err)
+			chkRow := chk.GetRow(0)
+			cdt := chkRow.GetDatumRow(fts)
+			for i, td := range testCase.data {
+				d := cdt[i]
+				if d.Kind() == types.KindMysqlDecimal {
+					require.Equal(t, td.bt.GetMysqlDecimal(), d.GetMysqlDecimal())
+				} else {
+					require.Equal(t, td.dt, d)
+				}
 			}
-			d := cdt[i]
-			c.Assert(d.IsNull(), Equals, true)
-		}
 
-		// decode to old row bytes.
-		colOffset := make(map[int64]int)
-		for i, t := range testData {
-			colOffset[t.id] = i
-		}
-		bDecoder := rowcodec.NewByteDecoder(cols, []int64{-1}, bdf, sc.TimeZone)
-		oldRow, err := bDecoder.DecodeToBytes(colOffset, kv.IntHandle(-1), newRow, nil)
-		c.Assert(err, IsNil)
-		for i, t := range testData {
-			remain, d, err := codec.DecodeOne(oldRow[i])
-			c.Assert(err, IsNil)
-			c.Assert(len(remain), Equals, 0)
-			if d.Kind() == types.KindMysqlDecimal {
-				c.Assert(d.GetMysqlDecimal(), DeepEquals, t.bt.GetMysqlDecimal())
-			} else {
-				c.Assert(d, DeepEquals, t.bt)
+			// decode to old row bytes.
+			colOffset := make(map[int64]int)
+			for i, td := range testCase.data {
+				colOffset[td.id] = i
 			}
-		}
+			bDecoder := rowcodec.NewByteDecoder(cols, []int64{-1}, nil, nil)
+			oldRow, err := bDecoder.DecodeToBytes(colOffset, kv.IntHandle(-1), newRow, nil)
+			require.NoError(t, err)
+			for i, td := range testCase.data {
+				remain, d, err := codec.DecodeOne(oldRow[i])
+				require.NoError(t, err)
+				require.Equal(t, 0, len(remain))
+				if d.Kind() == types.KindMysqlDecimal {
+					require.Equal(t, td.bt.GetMysqlDecimal(), d.GetMysqlDecimal())
+				} else if d.Kind() == types.KindBytes {
+					require.Equal(t, td.bt.GetBytes(), d.GetBytes())
+				} else {
+					require.Equal(t, td.bt, d)
+				}
+			}
+		})
 	}
-	dtNilData := []testData{
+}
+
+func TestNilAndDefault(t *testing.T) {
```

Comment:
Maybe add this test as a subtest into `TestTypesNewRowCodec`. The diff of this PR is hard to review. Inline expands the code too much.

In such case, I would like to see reusing as much as possible, e.g. merging two tests of `encodeAndDecode` into one.